src/main/java/org/opensearch/ad/task/ADTaskManager.java

```diff
@@ -1538,6 +1544,59 @@ public void updateLatestRealtimeTask(
         }, null, false, listener);
     }
 
+    /**
+     * Update latest realtime task's state, init progress, estimated left minutes and error field.
+     * @param detectorId detector id
+     * @param newState new task state which will override state calculated with rcf total updates
+     * @param rcfTotalUpdates total updates of RCF model
+     * @param detectorIntervalInMinutes detector interval in minutes
+     * @param error error message
+     */
+    public void updateLatestRealtimeTask(
+        String detectorId,
+        String newState,
+        Long rcfTotalUpdates,
+        Long detectorIntervalInMinutes,
+        String error
+    ) {
+        Float initProgress = null;
+        String state = null;
+        // calculate init progress and task state with RCF total updates
+        if (detectorIntervalInMinutes != null && rcfTotalUpdates != null) {
+            state = ADTaskState.INIT.name();
+            if (rcfTotalUpdates < NUM_MIN_SAMPLES) {
+                initProgress = (float) rcfTotalUpdates / NUM_MIN_SAMPLES;
+            } else {
+                state = ADTaskState.RUNNING.name();
+                initProgress = 1.0f;
+            }
+        }
+        // Check if new state is not null and override state calculated with rcf total updates
+        if (newState != null) {
+            state = newState;
+        }
+
+        error = Optional.ofNullable(error).orElse("");
+        if (!adTaskCacheManager.realtimeTaskChanged(detectorId, state, initProgress, error)) {
+            return;
+        }
+        if (ADTaskState.STOPPED.name().equals(state)) {
+            adTaskCacheManager.removeRealtimeTaskCache(detectorId);
```

Comment:
I only see a detector's removed from the cache if it is stopped or deleted. What about the failed state?

Reply:
If one interval failed, the realtime job will not stop, so we will keep using current task

Reply:
Got it. So any prolonged failure that leads to an EndRunException will lead to stopped state, which will remove the task from the cache. Makes sense